internal/sys/file_table.go

```diff
@@ -0,0 +1,141 @@
+package sys
+
+import "math/bits"
+
+// FileTable is a data structure mapping 32 bit file descriptor to file objects.
+//
+// Key generation is managed by the table, and currently uses a strategy similar
+// to the file descriptor allocation on unix systems: the lowest key not mapped
+// to any value is used when inserting a new item.
+//
+// The data structure is used to implement the file descriptor table of the WASI
+// implementation of Wazero. The WASI standard documents that programs cannot
+// expect that file descriptor numbers will be allocated with a lowest-first
+// strategy (like it is done on unix systems), and they shouldi instead assume
+// the values will be random. Since "random" is a very imprecise concept in
+// computers, we technically satisfying the implementation with the key
+// allocation strategy we use here. We could imagine adding more "randomness"
+// to the key selection process, however this should never be used as a security
+// measure to prevent applications from guessing the next file number so there
+// are no strong incentives to complicate the logic.
+//
+// The data structure optimizes for memory density and lookup performance,
+// trading off compute at insertion time. This is a useful compromise for the
+// use cases we employ it with: files are usually read or written a lot more
+// often than they are opened, each operation requires a table lookup so we are
+// better off spending extra compute to insert files in the table in order to
+// get cheaper lookups. Memory efficiency is also crucial to support scaling
+// with programs that open thousands of files: having a high or non-linear
+// memory-to-item ratio could otherwise be used as an attack vector by malicous
+// applications attempting to damage performance of the host.
+type FileTable struct {
+	masks []uint64
+	files []*FileEntry
+}
+
+// Len returns the number of files stored in the table.
+func (t *FileTable) Len() (n int) {
+	// We could make this a O(1) operation if we cached the number of files in
+	// the table. More state usually means more problems, so until we have a
+	// clear need for this, the simple implementation may be a better trade off.
+	for _, mask := range t.masks {
+		n += bits.OnesCount64(mask)
+	}
+	return n
+}
+
+// Grow ensures that t has enough room for n files, potentially reallocating the
+// internal buffers if their capacity was too small to hold this many files.
+func (t *FileTable) Grow(n int) {
+	if n = (n*8 + 7) / 8; n > len(t.masks) {
```

Comment:
since this file looks benchmark optimized you can consider shifting vs multiplying by powers of 2

e.g. `n << 3 // n * 8`

Reply:
The Go compiler will do arithmetic operations already (e.g. n*8 is translated to n<<3). The `Grow` method is not intended to be called frequently since the capacity grows exponentially, its cost is amortized over repeated calls to `Insert`.

Unless we have a performance profile showing the optimization would be worth making I'd rather keep the arithmetic in a form that is more readable to human maintainers if that's OK with you.

Reply:
this was more a nit as I noticed you had some benchmarks. there are other code where we found this improved the results, but agree that "grow" is indeed not one of those frequently called things and thanks for pointing it out!